game-core/src/main/java/games/strategy/triplea/delegate/battle/TargetGroup.java

```diff
@@ -0,0 +1,90 @@
+package games.strategy.triplea.delegate.battle;
+
+import com.google.common.collect.Sets;
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.data.UnitType;
+import games.strategy.triplea.attachments.UnitAttachment;
+import games.strategy.triplea.delegate.Matches;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import lombok.AccessLevel;
+import lombok.Getter;
+import org.triplea.java.collections.CollectionUtils;
+
+/** Group of firing units and their targets. */
+@Getter(AccessLevel.PRIVATE)
+public class TargetGroup {
+
+  private final Set<UnitType> firingUnitTypes;
+  private final Set<UnitType> targetUnitTypes;
+
+  public TargetGroup(final UnitType firingUnitType, final Set<UnitType> targetUnitTypes) {
+    firingUnitTypes = Sets.newHashSet(firingUnitType);
+    this.targetUnitTypes = targetUnitTypes;
+  }
+
+  public Collection<Unit> getFiringUnits(final Collection<Unit> units) {
+    return CollectionUtils.getMatches(units, Matches.unitIsOfTypes(firingUnitTypes));
+  }
+
+  public Collection<Unit> getTargetUnits(final Collection<Unit> units) {
+    return CollectionUtils.getMatches(units, Matches.unitIsOfTypes(targetUnitTypes));
+  }
+
+  /**
+   * Find all target groupings for firing units and enemy units based on canNotTarget and
+   * canNotBeTargetedBy attributes. Also takes into account if any destroyers are present which
+   * cancel canNotBeTargetedBy. Sort all the target groups so those with the least number of targets
+   * appear first.
+   */
+  public static List<TargetGroup> newTargetGroups(
+      final Collection<Unit> units, final Collection<Unit> enemyUnits) {
+
+    final Set<UnitType> unitTypes =
+        units.stream().map(unit -> unit.getType()).collect(Collectors.toSet());
+    final Set<UnitType> enemyUnitTypes =
+        enemyUnits.stream().map(unit -> unit.getType()).collect(Collectors.toSet());
+    final List<TargetGroup> targetGroups = new ArrayList<TargetGroup>();
+    for (final UnitType unitType : unitTypes) {
+      final Set<UnitType> targets = findTargets(unitType, unitTypes, enemyUnitTypes);
+      addToTargetGroups(unitType, targets, targetGroups);
+    }
+    return sortTargetGroups(targetGroups);
+  }
+
+  private static Set<UnitType> findTargets(
+      final UnitType unitType, final Set<UnitType> unitTypes, final Set<UnitType> enemyUnitTypes) {
+    final Set<UnitType> targets = new HashSet<>(enemyUnitTypes);
+    targets.removeAll(UnitAttachment.get(unitType).getCanNotTarget());
+    return unitTypes.stream().anyMatch(Matches.unitTypeIsDestroyer())
+        ? targets
+        : targets.stream()
+            .filter(
+                target -> !UnitAttachment.get(target).getCanNotBeTargetedBy().contains(unitType))
+            .collect(Collectors.toSet());
+  }
+
+  private static void addToTargetGroups(
```

Comment:
nit, `addToTargetGroups` could be a bit more descriptive. Notably, I'm curious/not sure what exactly we are adding to target groups. I'm not honestly 100% sure how to describe what we're adding to target groups here, sorry to not have good suggestions for you on-hand.

Second, bit of a side note, if you're able to see a way, would be nice as well to be able to return a result rather than modifying the input param `targetGroups`.  In which case up on line 55 we could have something like:
```
targetGroups.addEligibleTargets(unitType, targets);
```

Reply:
I can't think of any better name. Essentially its adding the unit type and targets to either an exist target group or adding a new target group.

A method call like that wouldn't be possible unless you made TargetGroups a separate class instead of a list.